File: 1484-linked-list-in-binary-tree/1484-linked-list-in-binary-tree.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private val pattern: MutableList<Int?> = mutableListOf();
    private val prefixTab: MutableList<Int> = mutableListOf(0);
    private fun searchTree(root: TreeNode?, patternIdx: Int): Boolean {
        if (root == null) return false;
        var idx: Int = patternIdx;
        while (idx > 0 && root?.`val` != pattern[idx]) idx = prefixTab[idx - 1];
        idx += if (root?.`val` == pattern[idx]) 1 else 0;

        if (idx == pattern.size) return true;

        return searchTree(root?.left, idx) || searchTree(root?.right, idx);
    }
    fun isSubPath(head: ListNode?, root: TreeNode?): Boolean {
        pattern.add(head?.`val`);
        var patternIdx: Int = 0;
        var dh: ListNode? = head?.next;

        while (dh != null) {
            while (patternIdx > 0 && dh?.`val` != pattern[patternIdx]) patternIdx = prefixTab[patternIdx - 1];
            patternIdx += if (dh?.`val` == pattern[patternIdx]) 1 else 0;
            pattern.add(dh?.`val`);
            prefixTab.add(patternIdx);
            dh = dh?.next;
        }

        return searchTree(root, 0);
    }
}

File: 1484-linked-list-in-binary-tree/1484-linked-list-in-binary-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#pragma GCC optimize(O3, "unroll-loops")
#pragma GCC target ("avx")
#pragma GCC target ("-fsplit-loops")

class Solution {
private:
    bool searchTree(TreeNode* root, ListNode* head) {
        if (!head) return true;
        if (!root || (root->val != head->val)) return false;
        return searchTree(root->left, head->next) || searchTree(root->right, head->next);
    }
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (!root) return false;

        return searchTree(root, head) || isSubPath(head, root->left) || isSubPath(head, root->right);
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();